Ghost roles

Ultradragon005 opened the pull request:
- [x] remove unneccesary strings.
- [x] Fix evilspirit 
- [X] test, works with players?
- [X] Add to aftermeetingtasks (Added to Exiled Patch)
- [X] Ignore ghostrole in gameresults.

Ultradragon005 commented on Patches/PlayerControlPatch.cs:
```diff
@@ -1310,6 +1310,14 @@ public static void Postfix(PlayerControl __instance, [HarmonyArgument(0)] Player
+            target.RpcSetRole(RoleTypes.GuardianAngel);
+            target.RpcSetCustomRole(CustomRoles.GuardianAngelTOHE);
+        }
+            
```
Will add a crewmate_check once I'm done with properly setting it up (Ghost roles will be made in such a way that the first x crewmates to die get them)

Ultradragon005 replied:
And also neutrals I think, unless I can change impostor team to make guardian angel impostor compatible.

Ultradragon005 commented on Modules/OptionHolder.cs:
```diff
@@ -3821,6 +3835,38 @@ public static void SetupRoleOptions(int id, TabGroup tab, CustomRoles role, Cust
+
+        //CustomRoleSpawnChances.Add(role, spawnOption);
+        CustomRoleCounts.Add(role, countOption);
+        //if (role.RoleExist())
```
So basically find a check for https://i.gyazo.com/cad823b1cb3a003066eaafbc690ce7a6.gif here.

Ultradragon005 replied:
I might just make another string option which is "Enabled - Disabled" 😂

Ultradragon005 commented:
![image](https://github.com/0xDrMoe/TownofHost-Enhanced/assets/143632280/d2b092da-f251-4228-b91b-a93eb63c2e36)
okay, I know what to do now I think. I will probably make it work tomorrow

NikoCat233 commented:
Patch check protect to accept calls from GuardianAnge

Ultradragon005 commented:
time to find all the instances add-ons prevent spawning <:hehe_sob:1135754965260972123>

Ultradragon005 commented:
okay.... so RoleExist() probably won't work. so I will have make a custom IsEnable method for ghost using Option.GhostAssign()

Ultradragon005 commented on Patches/PlayerControlPatch.cs:
```diff
@@ -1310,6 +1316,40 @@ public static void Postfix(PlayerControl __instance, [HarmonyArgument(0)] Player
+            }
+
+        }
+
```
If this will be causing issues, I will make a custom harmony checkdeath prefix patch

Ultradragon005 commented:
semi works now (3 hours troubleshooting just to find out I had to use rpcsetrole<:hehe_sob:1135754965260972123> well... good thing I atleast found rpcset role harmony patch bcuz of that to cancel it setting to crewmateghost)

NikoCat233 commented:
bro you can always ask...

Ultradragon005 commented on Modules/OptionHolder.cs:
```diff
@@ -3821,6 +3847,43 @@ public static void SetupRoleOptions(int id, TabGroup tab, CustomRoles role, Cust
+        CustomRoleCounts.Add(role, countOption);
+        GetRoleSpawnMode(role);
+        
+        if (trueghostenable) 
```
Need help here, this dosen't work for some reason... I need help fidning a way to check if the setting is (0 or 1) before it get sent to all the other places (to prevent ghost-role from spawning)

Ultradragon005 replied:
refer to "GetRoleSpawnMode" method.

Ultradragon005 replied:
nvm, I think Ik the problem, it only runs after game starts I think

Ultradragon005 replied:
okay... so ongamestart patch I go trough a foreach (var plrrole in customrolescounts) which insttantly leaves if not ghost role. And this finally made it possible to detect on/off option.

LMK if you have any better solution (PS: doing almost anything in optionholder fucks shit up. and RandomSpawnPatch is run before Load() so I can't use setupghostroleoptions)

NikoCat233 replied:
Sorry but why do you need a "setupghostroleoptions"

Ultradragon005 replied:
> Sorry but why do you need a "setupghostroleoptions"

To be able and add them to the tab, and disable them from spawning normally, ghost roles will only spawn after a player has died. (Similar to setupadtroles, but I coulnd't quite find out how that disables spawing. So I found another method to do it)

Ultradragon005 replied:
wait... I'm gonna look at how the code looks now, you maybe right and it's uneccessaary, but I think not because I wouldn't use setuadtroleoption, but I'll see...

NikoCat233 commented:
You can just add a isGhostRole in Custom role helper
Use isghostrole=> continue in custom role selector
And assign the ghost role in murder player patch

Ultradragon005 commented:
> You can just add a isGhostRole in Custom role helper Use isghostrole=> continue in custom role selector And assign the ghost role in murder player patch

okay, I will look into it.

Ultradragon005 commented:
OHHH, I didn't know Continue; skips the foreach iteration, alrright the code will be much simpler after I make some changes 😅

Ultradragon005 commented:
about 95% done, just gotta remove unneccesary code I accidentaly added from my previous attempts. And fix evilspirit.

And test with players (& also see if modded clients needs extra rpc's)

Ultradragon005 commented:
@Tommy-XL all works! Also I found out impostor/neutral ghost roles are also possible, but first let's test how Warden works, incase there's any bugs, but so far in my test, there was none.

Ultradragon005 commented:
I will add setupsingleoptionroles in a pr after this has already been merged.

Ultradragon005 commented:
Summary:
- Added ghost roles functionaly.
- Added Crewmate [Ghost] role Warden.
- Fixed Spiritcaller (haven't moved it out of experimental yet, bcuz the code of it was never tested in the first place 💀)

Tommy-XL commented:
I think that merge is after the release of 1.5.0
Because we need to focus on fixing bugs rather than adding something new

Tommy-XL commented:
Fix conflicts pls

Tommy-XL commented:
Can you create a Ghost folder and create two more Impostor/Crewmate folders there?

![image](https://github.com/0xDrMoe/TownofHost-Enhanced/assets/104814436/5df1fe98-dbb5-4cff-a12f-f0727efef998)

Ultradragon005 commented:
@Tommy-XL done!

Tommy-XL commented:
Like here

![image](https://github.com/0xDrMoe/TownofHost-Enhanced/assets/104814436/2aabfb55-8e97-427f-909f-7b302b44b98c)

Ultradragon005 commented:
> Like here
> 
> ![image](https://private-user-images.githubusercontent.com/104814436/302975253-2aabfb55-8e97-427f-909f-7b302b44b98c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDczMDg0ODAsIm5iZiI6MTcwNzMwODE4MCwicGF0aCI6Ii8xMDQ4MTQ0MzYvMzAyOTc1MjUzLTJhYWJmYjU1LThlOTctNDI3Zi05MDlmLTdiMzAyYjQ0Yjk4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwN1QxMjE2MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYTY0MDA1NDMyYjI3Y2IzZGJhZWQwNGNjYmI1M2Y0OTRiNjM3ZmJlZTE2MjAwYTM0MTVjNjJiNjE4OWJmZjQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.qj8xiEDziF2YCoqszfeOp0GDXLNGj4fh4OQj1bHKJ0I)

oh wait I knwo what you mean now.... Ig I will test that, I haven't tested that specific point, since the thing for evilspirit didn't seem for me to work anyways... so ignored it... but I will test

Ultradragon005 commented:
@Tommy-XL I change rpcsetrole patch to be the old one, altough I did not find any issues, hence it was there, I basically replace .is(evilspirit) with ALL ghost roles. So now it is fine.

Ultradragon005 commented:
so, is that all?
